src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -1790,7 +1806,7 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
         const auto & func = desc.function;
         std::optional<UInt64> num_rows{};
 
-        if (!query.prewhere() && !query.where())
+        if (!query.prewhere() && !query.where() && !context->getCurrentTransaction())
```

Comment:
It's because MergeTree maintains total count for all parts in Active state and it simply returns that number for trivial `select count() from table` query. But if we have current transaction, then we should return number of rows in current snapshot (that may include parts in Outdated state), so we have to use `totalRowsByPartitionPredicate` instead of `totalRows` even for trivial query.